File: articles/development-differentiation-making-of-organisms/fortran/development_kernel.f90
! Compact developmental growth and differentiation kernel in Fortran.

program development_kernel
  implicit none

  real :: N0
  real :: N24
  real :: r
  real :: doubling_time
  real :: K
  real :: dN
  real :: N
  real :: dt
  real :: k1
  real :: k2
  real :: lineage1_fraction
  real :: x
  real :: morphogen

  N0 = 1.0e4
  N24 = 4.0e4
  r = log(N24 / N0) / 24.0
  doubling_time = log(2.0) / r

  print *, "Growth rate:", r
  print *, "Doubling time:", doubling_time

  N = 1.0e4
  K = 6.2e4
  dt = 1.0
  dN = r * N * (1.0 - N / K)
  N = N + dN * dt

  print *, "Logistic next N:", N

  k1 = 0.14
  k2 = 0.09
  lineage1_fraction = k1 / (k1 + k2)

  print *, "Lineage 1 fraction:", lineage1_fraction

  x = 0.2
  morphogen = exp(-5.0 * x)

  print *, "Morphogen at x=0.2:", morphogen
end program development_kernel

File: articles/development-differentiation-making-of-organisms/c/development_kernel.c
/*
 * Compact developmental growth and differentiation kernel in C.
 */

#include <math.h>
#include <stdio.h>

double exponential_growth(double N0, double r, double t) {
    return N0 * exp(r * t);
}

double doubling_time(double r) {
    return log(2.0) / r;
}

double logistic_step(double N, double r, double K, double dt) {
    double dN = r * N * (1.0 - N / K);
    return N + dN * dt;
}

double morphogen_concentration(double x) {
    return exp(-5.0 * x);
}

int main(void) {
    double N0 = 1.0e4;
    double N24 = 4.0e4;
    double r = log(N24 / N0) / 24.0;

    printf("growth_rate=%.6f\n", r);
    printf("doubling_time_h=%.6f\n", doubling_time(r));
    printf("N_at_24h=%.2f\n", exponential_growth(N0, r, 24.0));
    printf("logistic_next=%.2f\n", logistic_step(1.0e4, 0.07, 6.2e4, 1.0));
    printf("morphogen_at_0_2=%.6f\n", morphogen_concentration(0.2));
    printf("lineage_1_fraction=%.6f\n", 0.14 / (0.14 + 0.09));

    return 0;
}
